cs3api4lab/auth/check_auth_interceptor.py

```diff
@@ -0,0 +1,39 @@
+import http
+
+import cs3
+import grpc
+import cs3.rpc.code_pb2 as cs3code
+
+
+class CheckAuthInterceptor(grpc.UnaryUnaryClientInterceptor,
+                           grpc.UnaryStreamClientInterceptor,
+                           grpc.StreamUnaryClientInterceptor,
+                           grpc.StreamStreamClientInterceptor):
+    unauth_codes = {cs3code.CODE_UNAUTHENTICATED}
+
+    def __init__(self, log, authenticator):
+        self.log = log
+        self.authenticator = authenticator
+
+    def intercept_unary_unary(self, continuation, client_call_details, request):
+        return self._intercept_call(continuation, client_call_details, request)
+
+    def intercept_stream_unary(self, continuation, client_call_details, request_iterator):
+        return self._intercept_call(continuation, client_call_details, request_iterator)
+
+    def intercept_unary_stream(self, continuation, client_call_details, request):
+        return self._intercept_call(continuation, client_call_details, request)
+
+    def intercept_stream_stream(self, continuation, client_call_details, request_iterator):
+        return self._intercept_call(continuation, client_call_details, request_iterator)
+
+    def _intercept_call(self, continuation, client_call_details, request_or_iterator):
```

Comment:
I'm not sure how this works. Please provide info.

But.. I would intercept calls to REVA and:
1. check if the current token is still valid (by checking the expiration), if not refresh it. Only after that I would call the function we're wrapping.
2. if the call to reva failed due to auth problems, we refresh the token AND try again. This would allow us to try one more time and thus avoid having to check this in other parts of the code (or worse, sending the error to the ui).

Reply:
The interceptor is a second layer of checking auth token, first is in the specific authentication class. 
For examples: 

- RevaPassword doesn't check auth token because getting expiration time form config file (like a WOPI example) and refreshes token if the expiration time is over
- RevaToken checking expiration time before send a request to the Riva, but can't refresh token and throw Http 401 error if a token is expired
-  EOS type now doesn't check expiration time and refresh token, but can when I will finish the mock server and create code to refresh token (next PR)

In future, we need to add code to resend request if a token is a refresh, but it's not trivial because we will have some probability of conflict situations. (New issue ?)

Reply:
I still don't understand why you need a mock server for EOS (unless you're trying to do something else).. This has nothing to do with EOS, is generic oAuth. The only thing specific to EOS is the file contents format which has other elements besides the banal token.

I can create an issue with the retry. But I don't understand why we have 2 layers of checks. Having 2 layers makes the retrial in the interceptor unnecessary because the token won't be refreshed here.

There's something in this logic that I think is not going well... As I suggested from the beginning, we should call the check/refresh token in the authenticator, and that's it. If it fails, we return immediately a 401. If it needs to renew, we renew.
Since these authenticators should inherit from a generic class that does a "login" agains REVA, it means that the check will do the following:

1. Is the reva token ok? if so, return
1. if oauth
    1. is the oauth token valid? If so, call login again and return new token
    1. else raise 401
1. if user/pass
    1.  call login again and return new token

Reply:
If eos token is a similar oAuth token I'm using mock to creating /refreshing oAuth token because: 
- I'm don't have a test system with the OAuth server like Keycloak. 
- Configure Keycloakor or another service is something like "cannon on flies". 
- In future, we can use a mock server to mock other users case. 
- Example of "jupyterhub/oauthenticator" mock all request/response to external services.

Interceptor is a good way to check general errors in one place (now for an example 401) whit SOLID, DRY rule. Now isn't general checker because expiry time in the token is checked before request.

This is slowly turning to an architectural problem :). How much business logic parent class must have? For me, it's the parent class is thin it's better. Each specialized code I would like to move to inherit classes.

Ok, I will prepare the auth classes with these rules.

Reply:
(Just saying again, there is not EOS token in this context, just oAuth written to a file.)

I like interceptors, I just don't see any utility at the moment in this implementation. Checking the validity of the tokens should be done before, in a place where we can act if there was any expiration. 

I understand and agree that the parent class should be thin. But, at the moment the authentication uses always the same logic, which is doing a `AuthenticateRequest` to REVA with 3 parameters. *Always*. If you put this logic in the children, then you have a lot of duplication.
In the future, we might have different methods to authenticate to REVA, but right now we only have one.

Reply:
In the last commit, I'm refactored and moved the basic auth method to the parent class.